gradle diktatFix task cause internal error

## Describe the bug

`diktatFix` task cause following error:

```
[ERROR] 2022-01-31 23:14:36 Internal error has occurred in rule [kdoc-methods]. Please make an issue on this bug at https://github.com/cqfn/diKTat/. 
                       As a workaround you can disable these inspections in yml config: <[KDOC_TRIVIAL_KDOC_ON_FUNCTION, KDOC_WITHOUT_PARAM_TAG, KDOC_WITHOUT_RETURN_TAG, KDOC_WITHOUT_THROWS_TAG, MISSING_KDOC_ON_FUNCTION]>.
                       Root cause of the problem is in [/home/yewton/Projects/testmmdiktat/app/src/main/kotlin/net/yewton/testmmdiktat/app/MessageUtils.kt] file.
java.lang.ExceptionInInitializerError: null
```

```
[ERROR] 2022-01-31 23:14:36 Internal error has occurred in rule [file-structure]. Please make an issue on this bug at https://github.com/cqfn/diKTat/. 
                       As a workaround you can disable these inspections in yml config: <[FILE_CONTAINS_ONLY_COMMENTS, FILE_INCORRECT_BLOCKS_ORDER, FILE_NO_BLANK_LINE_BETWEEN_BLOCKS, FILE_UNORDERED_IMPORTS, FILE_WILDCARD_IMPORTS, UNUSED_IMPORT]>.
                       Root cause of the problem is in [/home/yewton/Projects/testmmdiktat/app/src/test/kotlin/net/yewton/testmmdiktat/app/MessageUtilsTest.kt] file.
java.lang.NoClassDefFoundError: Could not initialize class org.jetbrains.kotlin.com.intellij.openapi.util.objectTree.ThrowableInterner
```

## Expected behavior

It fixes without error

## Observed behavior

Described error occurs

## Steps to Reproduce

Checkout https://github.com/yewton/testmmdiktat and run `./gradlew diktatFix`

This project is generated by `gradle init` -> type: application -> lang: Kotlin -> multiple projects: yes

## Environment information
* diktat version:1.0.2
* build tool (maven/gradle):gralde
* how is diktat run (CLI, plugin, etc.):CLI
* kotlin version: ~1.5.31~ *1.6.10*
* operating system:Xubuntu Jammy Jellyfish (development branch) x86_64
* link to a project (if your project is public): https://github.com/yewton/testmmdiktat

--- a/diktat-gradle-plugin/src/main/kotlin/org/cqfn/diktat/plugin/gradle/DiktatGradlePlugin.kt
+++ b/diktat-gradle-plugin/src/main/kotlin/org/cqfn/diktat/plugin/gradle/DiktatGradlePlugin.kt
@@ -69,5 +69,10 @@ class DiktatGradlePlugin : Plugin<Project> {
          * Task to run diKTat with fix
          */
         const val DIKTAT_FIX_TASK = "diktatFix"
+
+        /**
+         * Version of JVM with more strict module system, which requires `add-opens` for kotlin compiler
+         */
+        const val MIN_JVM_REQUIRES_ADD_OPENS = 16
     }
 }

--- a/diktat-gradle-plugin/src/main/kotlin/org/cqfn/diktat/plugin/gradle/DiktatJavaExecTaskBase.kt
+++ b/diktat-gradle-plugin/src/main/kotlin/org/cqfn/diktat/plugin/gradle/DiktatJavaExecTaskBase.kt
@@ -6,6 +6,7 @@ import org.cqfn.diktat.ruleset.rules.DIKTAT_CONF_PROPERTY
 
 import generated.DIKTAT_VERSION
 import generated.KTLINT_VERSION
+import org.gradle.api.JavaVersion
 import org.gradle.api.Project
 import org.gradle.api.artifacts.Configuration
 import org.gradle.api.file.FileCollection
@@ -33,7 +34,7 @@ import javax.inject.Inject
  * Note: class being `open` is required for gradle to create a task.
  */
 open class DiktatJavaExecTaskBase @Inject constructor(
-    gradleVersionString: String,
+    private val gradleVersionString: String,
     diktatExtension: DiktatExtension,
     diktatConfiguration: Configuration,
     private val inputs: PatternFilterable,
@@ -133,6 +134,7 @@ open class DiktatJavaExecTaskBase @Inject constructor(
      */
     @TaskAction
     override fun exec() {
+        fixForNewJpms()
         if (shouldRun) {
             super.exec()
         } else {
@@ -212,6 +214,27 @@ open class DiktatJavaExecTaskBase @Inject constructor(
                 firstOrNull { it.exists() } ?: first()
             }
             .absolutePath
+    }
+
+    private fun fixForNewJpms() {
+        val javaVersion = getJavaExecJvmVersion()
+        project.logger.debug("For diktat execution jvm version $javaVersion will be used")
+        if (javaVersion >= JavaVersion.VERSION_16) {
+            // https://github.com/analysis-dev/diktat/issues/1182#issuecomment-1023099713
+            project.logger.debug("Adding `--add-opens` flag for JVM version >=16 compatibility")
+            jvmArgs("--add-opens", "java.base/java.lang=ALL-UNNAMED")
+        }
+    }
+
+    private fun getJavaExecJvmVersion(): JavaVersion = if (GradleVersion.version(gradleVersionString) >= GradleVersion.version("6.7") &&
+            javaLauncher.isPresent
+    ) {
+        // Java Launchers are available since 6.7, but may not always be configured
+        javaLauncher.map { it.metadata.jvmVersion }.map(JavaVersion::toVersion).get()
+    } else {
+        // `javaVersion` property is available since 5.2 and is simply derived from path to JVM executable,
+        // which may be explicitly set or be the same as current Gradle JVM.
+        javaVersion
     }
 }
 

--- a/diktat-gradle-plugin/src/main/kotlin/org/cqfn/diktat/plugin/gradle/DiktatJavaExecTaskBase.kt
+++ b/diktat-gradle-plugin/src/main/kotlin/org/cqfn/diktat/plugin/gradle/DiktatJavaExecTaskBase.kt
@@ -2,6 +2,7 @@ package org.cqfn.diktat.plugin.gradle
 
 import org.cqfn.diktat.plugin.gradle.DiktatGradlePlugin.Companion.DIKTAT_CHECK_TASK
 import org.cqfn.diktat.plugin.gradle.DiktatGradlePlugin.Companion.DIKTAT_FIX_TASK
+import org.cqfn.diktat.plugin.gradle.DiktatGradlePlugin.Companion.MIN_JVM_REQUIRES_ADD_OPENS
 import org.cqfn.diktat.ruleset.rules.DIKTAT_CONF_PROPERTY
 
 import generated.DIKTAT_VERSION
@@ -219,9 +220,9 @@ open class DiktatJavaExecTaskBase @Inject constructor(
     private fun fixForNewJpms() {
         val javaVersion = getJavaExecJvmVersion()
         project.logger.debug("For diktat execution jvm version $javaVersion will be used")
-        if (javaVersion.majorVersion.toInt() >= 16) {
+        if (javaVersion.majorVersion.toInt() >= MIN_JVM_REQUIRES_ADD_OPENS) {
             // https://github.com/analysis-dev/diktat/issues/1182#issuecomment-1023099713
-            project.logger.debug("Adding `--add-opens` flag for JVM version >=16 compatibility")
+            project.logger.debug("Adding `--add-opens` flag for JVM version >=$MIN_JVM_REQUIRES_ADD_OPENS compatibility")
             jvmArgs("--add-opens", "java.base/java.lang=ALL-UNNAMED")
         }
     }

--- a/diktat-gradle-plugin/src/main/kotlin/org/cqfn/diktat/plugin/gradle/DiktatJavaExecTaskBase.kt
+++ b/diktat-gradle-plugin/src/main/kotlin/org/cqfn/diktat/plugin/gradle/DiktatJavaExecTaskBase.kt
@@ -219,7 +219,7 @@ open class DiktatJavaExecTaskBase @Inject constructor(
     private fun fixForNewJpms() {
         val javaVersion = getJavaExecJvmVersion()
         project.logger.debug("For diktat execution jvm version $javaVersion will be used")
-        if (javaVersion >= JavaVersion.VERSION_16) {
+        if (javaVersion.majorVersion.toInt() >= 16) {
             // https://github.com/analysis-dev/diktat/issues/1182#issuecomment-1023099713
             project.logger.debug("Adding `--add-opens` flag for JVM version >=16 compatibility")
             jvmArgs("--add-opens", "java.base/java.lang=ALL-UNNAMED")

